### DISCUSSION TITLE
Where is the sampling for SAGEConv implemented?

### DISCUSSION THREAD

**Discussion by OliEfr** [pyg-team/pytorch_geometric#9106] [Q&A] [answered]
Hi all!

According to the [paper](https://arxiv.org/abs/1706.02216) GraphSAGE (SAGEConv) implements a neighbourhood sampling for the nodes. I couldn`t find this implementation detail in the code for pyg. Could someone please hint me to the respective LOC in the code?

Thank you :)

**Comment by rusty1s** [comment]
We have seperated GNN layer and neighbor sampling routine in PyG:
* `SAGEConv` implements the GNN layer
* `NeighborLoader` implements the neighborhood sampling routine

**Reply by OliEfr** [reply]
Ok, so when I create a NN with a  `SAGEConv` layer the `NeighborLoader` is called somewhere during the forward pass (without the need for me to implement anything), correct? Could you point me to the LOC where this happaning? Sorry, I can not seem to find it (stepping through the code with the debugger also didn't help me).

A lot of thanks in advance, and sorry for my inexperience.

**Reply by OliEfr** [reply]
Ok I found [this example](https://github.com/pyg-team/pytorch_geometric/discussions/7371), which uses the `NeighborLoader` as the dataloader (as opposed to "normal" `DataLoader`). Is that what I need to do?

 However, I have been using the `SAGEConv` module in a heterogenous graph with the "normal" `DataLoader`. What would that mean for the sampling. Would just all neighbor nodes be sampled in that case?

**Reply by rusty1s** [reply, marked as answer]
> Ok I found https://github.com/pyg-team/pytorch_geometric/discussions/7371, which uses the NeighborLoader as the dataloader (as opposed to "normal" DataLoader). Is that what I need to do?

Yes, exactly.

> Would just all neighbor nodes be sampled in that case?

Yes, there is no sampling inside `SAGEConv`.

**Reply by OliEfr** [reply]
Ok thank you. Maybe its a good idea to add a hint in the documentation regarding that.

**Reply by arunraja-hub** [reply]
@rusty1s does that mean that just SAGEConv with its default parameters is just a vanilla message passing layer?
